linkerd/http-access-log/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+struct ResponseFutureInner {
+    span: Span,
+    start: Instant,
+    processing: Duration,
+}
+
+#[pin_project]
+pub struct AccessLogFuture<F> {
+    data: Option<ResponseFutureInner>,
+
+    #[pin]
+    inner: F,
+}
+
+impl<N> NewAccessLog<N> {
+    #[inline]
+    pub fn layer() -> impl svc::layer::Layer<N, Service = Self> {
+        svc::layer::mk(|inner| NewAccessLog { inner })
+    }
+}
+
+impl<N, T> NewService<T> for NewAccessLog<N>
+where
+    T: Param<tls::ConditionalServerTls> + Param<Remote<ClientAddr>>,
+    N: NewService<T>,
+{
+    type Service = AccessLogContext<N::Service>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let Remote(ClientAddr(client_addr)) = target.param();
+        let tls: tls::ConditionalServerTls = target.param();
+        let client_id = tls
+            .value()
+            .and_then(|tls| tls.client_id().map(|tls::ClientId(name)| name.clone()));
+        let inner = self.inner.new_service(target);
+        AccessLogContext {
+            inner,
+            client_addr,
+            client_id,
+        }
+    }
+}
+
+impl<S, B1, B2> svc::Service<http::Request<B1>> for AccessLogContext<S>
+where
+    S: svc::Service<http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = AccessLogFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: http::Request<B1>) -> Self::Future {
+        let get_header = |name: http::header::HeaderName| {
+            request
+                .headers()
+                .get(name)
+                .and_then(|x| x.to_str().ok())
+                .unwrap_or_default()
+        };
+
+        let trace_id = || {
+            let headers = request.headers();
+            headers
+                .get("x-b3-traceid")
+                .or_else(|| headers.get("x-request-id"))
+                .or_else(|| headers.get("x-amzn-trace-id"))
+                .and_then(|x| x.to_str().ok())
+                .unwrap_or_default()
+        };
+
+        let span = span!(target: TRACE_TARGET, Level::INFO, "http",
+            client.addr = %self.client_addr,
+            client.id = self.client_id.as_ref().map(|n| n.as_str()).unwrap_or("-"),
+            timestamp = %now(),
+            method = request.method().as_str(),
+            uri =  %request.uri(),
+            version = ?request.version(),
+            trace_id = trace_id(),
+            request_bytes = get_header(http::header::CONTENT_LENGTH),
+            status = field::Empty,
+            response_bytes = field::Empty,
+            total_ns = field::Empty,
+            processing_ns = field::Empty,
+            user_agent = get_header(http::header::USER_AGENT),
+            host = get_header(http::header::HOST),
+        );
+
+        if span.is_disabled() {
```

Comment:
I think it's worth leaving a comment here making it clear that the `span!` body is only materialized when access logging is enabled by `linkerd_tracing::access_log`'s initializer. I.e. this is expected to be free when tracing is not enabled.

Maybe also worth leaving a comment on the `fn layer` summarizing this, too.